File: dijkstra/dijkstra.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <utility>
#include <unordered_map>
#include <queue>
#include <sstream>

using namespace std;

typedef int To;
typedef int Cost;
typedef unordered_map<To,Cost> Edges;

class Node {
    int id;
    Edges edges;

    public:
        Node(int from):id(from){};

        void add_edge(int to_node, int cost){
            edges[to_node] = cost;
        }

        int get_id(){
            return id;
        }

        Edges* get_edges(){
            return &edges;
        }
};

typedef int Id;
typedef unordered_map<Id,Node*> Nodes;

class Graph {
    protected:
        Nodes nodes;

    public:
        Graph(){};
        void add_node(int from){
            if(nodes.find(from)==nodes.end())
                nodes[from] = new Node(from);
        }

        void add_edge(int from, int to, int cost){
            Node* node = nodes[from];
            node->add_edge(to,cost);
        }

        virtual void solve(int,int) = 0;
        virtual void show_path() = 0;
};


typedef struct ManageNode {
    Node* node;
    int cost;
    bool done;
    int from;
    ManageNode(Node *node, int cost): node(node), cost(cost), done(false), from(-1) {};
} ManageNode;

class compClass {
    public:
        bool operator() (ManageNode* a, ManageNode* b){
            return a->cost > b->cost;
        }
};

typedef unordered_map<Id,ManageNode*> ManageNodes;
typedef priority_queue<ManageNode*, vector<ManageNode*>, compClass> QueueNodes;

string reduce(vector<int>& path){
    stringstream ss;
    for(int i=0;i<path.size()-1;i++){
        ss << path[i] << " -> ";
    }
    ss << path[path.size()-1];
    return ss.str();
}

class Dijkstra : public Graph {
    ManageNodes man_nodes;
    QueueNodes pq_nodes;
    int start, goal;

    void init_man_nodes(){
        for(auto node_info: nodes){
            int node_id = node_info.first;
            Node* node = node_info.second;
            man_nodes[node_id] = new ManageNode(node,1<<30);
        }
    }

    void update_man_node(int node_id, int cost, int from){
        man_nodes[node_id]->cost = cost;
        man_nodes[node_id]->from = from;
    }

    void enque_man_node(int node_id){
        ManageNode* man_node = man_nodes[node_id];
        pq_nodes.push( man_node );
    }

    int get_man_node_cost(int node_id){
        return man_nodes[node_id]->cost;
    }

    Node* take_min_cost_node(){
        return pq_nodes.top()->node;
    }

    int take_min_cost(){
        return pq_nodes.top()->cost;
    }

    void remove_min_cost_node(){
        pq_nodes.pop();
    }

    bool judge_if_done(int node_id){
        return man_nodes[node_id]->done;
    }

    void save_done(int node_id){
        man_nodes[node_id]->done = true;
    }

    public:
        Dijkstra() {};

        void solve(int start, int goal){
            this->start = start;
            this->goal = goal;

            cout << start << " ---> " << goal << endl;

            init_man_nodes();
            update_man_node(start, 0, -1);
            enque_man_node(start);

            int to_goal_cost = 0;

            Node *node;
            int cost, node_id;
            while(!pq_nodes.empty()){
                node = take_min_cost_node();
                cost = take_min_cost();
                node_id = node->get_id();

                remove_min_cost_node();

                if(judge_if_done(node_id)){
                    continue;
                }else{
                    save_done(node_id);
                }

                if(node_id==goal){
                    to_goal_cost = cost;
                    break;
                }

                int to_node_id, to_node_cost, new_cost;
                for( auto edge: *(node->get_edges()) ){
                    to_node_id = edge.first;
                    to_node_cost = edge.second;

                    new_cost = cost+to_node_cost;
                    if(new_cost < get_man_node_cost(to_node_id)){
                        update_man_node(to_node_id, new_cost, node_id);
                        enque_man_node(to_node_id);
                    }
                }
            }
            cout << "cost : " << to_goal_cost << endl;
        }

        void show_path(){
            cout << "path : ";
            vector<int> path;
            int node_id = goal;
            for(;;){
                path.push_back(node_id);
                if(node_id==start)
                    break;
                node_id = man_nodes[node_id]->from;
            }
            reverse(path.begin(),path.end());
            cout << reduce(path) << endl;
        }
};

int main(int argc, char* argv[]) {

    int num_nodes, num_edges;
    int start_node, goal_node;

    Graph* graph = new Dijkstra();

    if(argc==2) {
        ifstream ifs(argv[1]);
        ifs >> num_nodes >> num_edges;
        ifs >> start_node >> goal_node;
    
        int from_node, to_node, edge_cost;
        while( ifs >> from_node >> to_node >> edge_cost ){
            graph->add_node(from_node);
            graph->add_node(to_node);
            graph->add_edge(from_node,to_node,edge_cost);
        }

    }else{
        cin >> num_nodes >> num_edges;
        cin >> start_node >> goal_node;

        int from_node, to_node, edge_cost;
        for(int i=0;i<num_edges;i++){
            cin >> from_node >> to_node >> edge_cost;
            graph->add_node(from_node);
            graph->add_node(to_node);
            graph->add_edge(from_node,to_node,edge_cost);
        }
    }

    graph->solve(start_node,goal_node);
    graph->show_path();

    return 0;
}

File: dijkstra/dijkstra.rs
use std::error::Error;
use std::env;
use std::fs::File;
use std::path::Path;
use std::io::prelude::*;
use std::collections::HashMap;
use std::io::BufReader;
use std::cmp::Ordering;
use std::collections::BinaryHeap;
use std::io;

type FromNode = i32;
type ToNode = i32;
type EdgeCost= i32;

#[derive(Clone,Eq,PartialEq)]
struct Node {
    id: i32,
    edges: HashMap<ToNode,EdgeCost>,
}

struct Graph {
    nodes: HashMap<FromNode,Box<Node>>,
}

impl Graph {
    fn add_node(&mut self,from_node:i32) {
        if !self.nodes.contains_key(&from_node) {
            let node = Box::new(Node{id:from_node, edges:HashMap::new()});
            self.nodes.insert(from_node,node); // move ownership of Box node?
        }
    }
   
    fn add_edge(&mut self,from_node:i32,to_node:i32,edge_cost:i32) {
        if self.nodes.contains_key(&from_node) {
            let node = self.nodes.get_mut(&from_node).unwrap();
            node.edges.insert(to_node,edge_cost);
        }
    }
}

#[derive(Clone,Eq,PartialEq)]
struct ManageNode {
    cost : i32,
    from : i32,
    done : bool,
    node : Box<Node>,
}

impl Ord for ManageNode {
    fn cmp(&self, other: &ManageNode) -> Ordering {
        other.cost.cmp(&self.cost)
    }
}

impl PartialOrd for ManageNode {
    fn partial_cmp(&self, other: &ManageNode) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

struct Dijkstra {
    start : i32,
    goal : i32,
    graph : Graph, // Composition
    man_nodes: HashMap<FromNode,ManageNode>,
    heap_nodes: BinaryHeap<Box<ManageNode>>,
}

impl Dijkstra {
    fn new(start:i32, goal:i32, graph:Graph) -> Dijkstra {
        Dijkstra {
            start: start,
            goal: goal,
            graph: graph,
            man_nodes: HashMap::new(),
            heap_nodes: BinaryHeap::new(),
        }
    }

    fn init_man_nodes(&mut self){
        for (id,node) in &self.graph.nodes {
            self.man_nodes.insert(*id,ManageNode{cost:i32::max_value(),from:-1,done:false,node:node.clone()});
        }
    }

    fn enque_man_node(&mut self, id:&i32, cost:i32, from:i32){
        let mut man_node = self.man_nodes.get_mut(id).unwrap();
        man_node.cost = cost;
        man_node.from = from;
        self.heap_nodes.push(Box::new((*man_node).clone()));
    }

    fn solve(&mut self){

        self.init_man_nodes();
        let node = self.graph.nodes.get(&self.start).unwrap().clone();
        self.enque_man_node(&node.id,0,-1);

        let mut to_goal_cost = 0;
        while let Some(mut man_node) = self.heap_nodes.pop() {
            if man_node.done { continue }
            man_node.done = true;
            if man_node.node.id == self.goal {
                to_goal_cost = man_node.cost;
                break;
            }
            for (to_node, edge_cost) in &man_node.node.edges {
                let new_cost = man_node.cost + edge_cost;
                if new_cost < self.man_nodes.get(&to_node).unwrap().cost {
                    self.man_nodes.get_mut(&to_node).unwrap().cost = new_cost;
                    self.enque_man_node(&to_node,new_cost,man_node.node.id);
                }
            }
        }
        println!("cost : {}",to_goal_cost);
    }

    fn show_path(&self){
        let mut path:Vec<i32> = Vec::new();
        let mut id = self.goal;
        while id != self.start {
            path.push(id);
            let man_node = self.man_nodes.get(&id).unwrap();
            id = man_node.from;
        }
        path.push(self.start);
        path.reverse();
        println!("path : {}",path.iter().map(|x|x.to_string()).collect::<Vec<String>>().join(" -> "));
    }
}

fn main(){
    let args: Vec<String> = env::args().collect();

    let mut graph = Graph{nodes:HashMap::new()};
    let mut start = 0;
    let mut goal = 0;

    if args.len() == 2 {

        let path = Path::new(&args[1]);
        let file = match File::open(&path) {
            Err(why) => panic!("Couldn't open {}",why.description()),
            Ok(file) => file,
        };
    
        let buf = BufReader::new(file);
        for (line,n) in buf.lines().zip(1..) {
            let input:Vec<i32> = line.unwrap().split_whitespace().map(|x|x.parse::<i32>().unwrap()).collect();
            if n == 1 {
                let _ = input[0];
                let _ = input[1];
            }else if n == 2 {
                start = input[0];
                goal = input[1];
            }else if n > 2 {
                let from_node = input[0];
                let to_node = input[1];
                let edge_cost = input[2];
                graph.add_node(from_node);
                graph.add_node(to_node);
                graph.add_edge(from_node,to_node,edge_cost);
            }
        }

    } else {

        let stdin = io::stdin();
        for (line,n) in stdin.lock().lines().zip(1..) {
            let input:Vec<i32> = line.unwrap().split_whitespace().map(|x|x.parse::<i32>().unwrap()).collect();
            if n == 1 {
                let _ = input[0];
                let _ = input[1];
            }else if n == 2 {
                start = input[0];
                goal = input[1];
            }else if n > 2 {
                let from_node = input[0];
                let to_node = input[1];
                let edge_cost = input[2];
                graph.add_node(from_node);
                graph.add_node(to_node);
                graph.add_edge(from_node,to_node,edge_cost);
            }
        }

    }

    let mut dijkstra = Dijkstra::new(start,goal,graph);
    dijkstra.solve();
    dijkstra.show_path();

}

